### PR TITLE
Performance optimizations (Shifu v2)

### PR DESCRIPTION
**Pull request by moshe-blox**

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -32,7 +35,7 @@ Build stage Docker image:
     - docker tag $IMAGE_NAME:$CI_BUILD_REF $DOCKER_REPO_INFRA_STAGE:$CI_BUILD_REF
     - $DOCKER_LOGIN_TO_INFRA_STAGE_REPO && docker push $DOCKER_REPO_INFRA_STAGE:$CI_BUILD_REF
   only:
-    - stage
+    - fix-nil-decideds
```
Comment (lior-blox) [inline .gitlab-ci.yml:38]: @moshe-blox change to stage

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -66,11 +69,13 @@ Deploy ssv nodes to blox-infra-stage cluster:
     - mv kubectl /usr/bin/
     - .k8/scripts/deploy-ssv-nodes-yamls-on-stage-k8s.sh $DOCKER_REPO_INFRA_STAGE $CI_BUILD_REF ssv $APP_REPLICAS_INFRA_STAGE blox-infra-stage kubernetes-admin@blox-infra stage.ssv.network $K8S_API_VERSION $STAGE_HEALTH_CHECK_IMAGE $SSV_NODES_CPU_LIMIT $SSV_NODES_MEM_LIMIT
     - .k8/scripts/deploy-ssv-node-v1-yamls-on-stage-k8s.sh $DOCKER_REPO_INFRA_STAGE $CI_BUILD_REF ssv $APP_REPLICAS_INFRA_STAGE blox-infra-stage kubernetes-admin@blox-infra stage.ssv.network $K8S_API_VERSION $STAGE_HEALTH_CHECK_IMAGE $SSV_NODES_CPU_LIMIT_V1 $SSV_NODES_MEM_LIMIT_V1
-    - .k8/scripts/deploy-ssv-node-v3-yamls-on-stage-k8s.sh $DOCKER_REPO_INFRA_STAGE $CI_BUILD_REF ssv $APP_REPLICAS_INFRA_STAGE blox-infra-stage kubernetes-admin@blox-infra stage.ssv.network $K8S_API_VERSION $STAGE_HEALTH_CHECK_IMAGE $SSV_NODES_CPU_LIMIT $SSV_NODES_MEM_LIMIT
+#    - .k8/scripts/deploy-ssv-node-v3-yamls-on-stage-k8s.sh $DOCKER_REPO_INFRA_STAGE $CI_BUILD_REF ssv $APP_REPLICAS_INFRA_STAGE blox-infra-stage kubernetes-admin@blox-infra stage.ssv.network $K8S_API_VERSION $STAGE_HEALTH_CHECK_IMAGE $SSV_NODES_CPU_LIMIT $SSV_NODES_MEM_LIMIT
```
Comment (lior-blox) [inline .gitlab-ci.yml:72]: uncomment

File: go.mod [hunk 3]
```diff
@@ -193,4 +197,4 @@ replace github.com/google/flatbuffers => github.com/google/flatbuffers v1.11.0
 
 replace github.com/dgraph-io/ristretto => github.com/dgraph-io/ristretto v0.1.1-0.20211108053508-297c39e6640f
 
-replace github.com/bloxapp/ssv-spec => github.com/olegshmuelov/ssv-spec v0.0.0-20230117095155-f5d26c0be261
+replace github.com/bloxapp/ssv-spec => github.com/moshe-blox/ssv-spec v0.0.0-20230129120505-9f75252e0424
```
Comment (lior-blox) [inline go.mod:200]: remove

File: network/p2p/p2p_pubsub.go [hunk 4]
```diff
@@ -204,33 +192,42 @@ func (n *p2pNetwork) handlePubsubMessages(topic string, msg *pubsub.Message) err
 	if ssvMsg == nil {
 		return errors.New("message was not decoded")
 	}
-	// logger = withIncomingMsgFields(logger, msg, ssvMsg)
-	// logger.Debug("incoming pubsub message", zap.String("topic", topic),
-	//	zap.String("msgType", message.MsgTypeToString(ssvMsg.MsgType)))
-	metricsRouterIncoming.WithLabelValues(ssvMsg.GetID().String(), message.MsgTypeToString(ssvMsg.MsgType)).Inc()
+
```
Comment (lior-blox) [inline network/p2p/p2p_pubsub.go:195]: remove

File: protocol/v2/qbft/instance/instance.go [hunk 5]
```diff
@@ -84,20 +88,35 @@ func (i *Instance) Start(value []byte, height specqbft.Height) {
 }
 
 func (i *Instance) Broadcast(msg *specqbft.SignedMessage) error {
-	byts, err := msg.Encode()
-	if err != nil {
-		return errors.Wrap(err, "could not encode message")
-	}
+	go func() {
```
Comment (lior-blox) [inline protocol/v2/qbft/instance/instance.go]: need to decide without breaking spec

File: protocol/v2/ssv/runner/attester.go [hunk 6]
```diff
@@ -133,7 +140,8 @@ func (r *AttesterRunner) ProcessPostConsensus(signedMsg *specssv.SignedPartialSi
 
 		r.logger.Debug("reconstructed partial signatures",
 			zap.Any("signers", getPostConsensusSigners(r.GetState(), root)),
-			zap.Int64("slot", int64(duty.Slot)))
+			zap.Int64("slot", int64(duty.Slot)),
```
Comment (lior-blox) [inline protocol/v2/ssv/runner/attester.go:143]: remove

Comment (lior-blox) [reply protocol/v2/ssv/runner/attester.go:143]: duration

File: protocol/v2/ssv/runner/attester.go [hunk 7]
```diff
@@ -144,13 +152,23 @@ func (r *AttesterRunner) ProcessPostConsensus(signedMsg *specssv.SignedPartialSi
...
 		// Submit it to the BN.
+		start = time.Now()
 		if err := r.beacon.SubmitAttestation(signedAtt); err != nil {
 			r.logger.Error("failed to submit attestation to Beacon node",
 				zap.Int64("slot", int64(duty.Slot)), zap.Error(err))
 			return errors.Wrap(err, "could not submit to Beacon chain reconstructed attestation")
 		}
 
-		r.logger.Debug("successfully submitted attestation", zap.Int64("slot", int64(duty.Slot)))
+		var consensusTime time.Duration
```
Comment (lior-blox) [inline protocol/v2/ssv/runner/attester.go:162]: revert

File: protocol/v2/ssv/runner/attester.go [hunk 8]
```diff
@@ -172,13 +190,21 @@ func (r *AttesterRunner) expectedPostConsensusRootsAndDomain() ([]ssz.HashRoot,
 // 3) Once consensus decides, sign partial attestation and broadcast
 // 4) collect 2f+1 partial sigs, reconstruct and broadcast valid attestation sig to the BN
 func (r *AttesterRunner) executeDuty(duty *spectypes.Duty) error {
-	// TODO - waitOneThirdOrValidBlock
+	// TODO: stop waiting 1/3rd inside Beacon and move this here V
+	duration := time.Second * time.Duration(uint64(duty.Slot)*uint64(spectypes.PraterNetwork.SlotDurationSec().Seconds()))
```
Comment (lior-blox) [inline protocol/v2/ssv/runner/attester.go:194]: remove

File: protocol/v2/ssv/runner/attester.go [hunk 9]
```diff
@@ -172,13 +190,21 @@ func (r *AttesterRunner) expectedPostConsensusRootsAndDomain() ([]ssz.HashRoot,
...
+	start := time.Unix(int64(spectypes.PraterNetwork.MinGenesisTime()), 0).Add(duration).Add(4 * time.Second)
 
 	attData, err := r.GetBeaconNode().GetAttestationData(duty.Slot, duty.CommitteeIndex)
 	if err != nil {
 		return errors.Wrap(err, "failed to get attestation data")
 	}
 
+	// Log if fetching AttestationData took too long.
```
Comment (lior-blox) [inline protocol/v2/ssv/runner/attester.go:202]: revert

File: protocol/v2/ssv/validator/msgqueue_consumer.go [hunk 10]
```diff
@@ -38,8 +39,12 @@ func (v *Validator) HandleMessage(msg *spectypes.SSVMessage) {
 			)
 			return
 		}
-		//v.logMsg(decodedMsg, "adding to q", zap.Any("type", decodedMsg.SSVMessage.MsgType))
+		start := time.Now()
 		q.Q.Push(decodedMsg)
+		v.logMsg(decodedMsg, "pushed msg to queue",
```
Comment (lior-blox) [inline protocol/v2/ssv/validator/msgqueue_consumer.go:44]: remove

File: protocol/v2/ssv/validator/msgqueue_consumer.go [hunk 11]
```diff
@@ -51,7 +56,7 @@ func (v *Validator) StartQueueConsumer(msgID spectypes.MessageID, handler Messag
 	defer cancel()
 
 	for ctx.Err() == nil {
-		err := v.ConsumeQueue(msgID, handler, time.Millisecond*50)
+		err := v.ConsumeQueue(msgID, handler, time.Millisecond*15)
```
Comment (lior-blox) [inline protocol/v2/ssv/validator/msgqueue_consumer.go:59]: remove the timer

File: protocol/v2/ssv/validator/msgqueue_consumer.go [hunk 12]
```diff
@@ -91,14 +92,15 @@ func (v *Validator) ConsumeQueue(msgID spectypes.MessageID, handler MessageHandl
...
 
 		// Pop the highest priority message and handle it.
-		msg := q.Q.Pop(queue.NewMessagePrioritizer(&state))
-		if msg == nil {
-			//logger.Debug("could not pop message from queue")
-			time.Sleep(interval)
-			continue
-		}
+		msg := q.Q.WaitAndPop(queue.NewMessagePrioritizer(&state))
+		// if msg == nil {
```
Comment (lior-blox) [inline protocol/v2/ssv/validator/msgqueue_consumer.go:99]: remove comment

File: protocol/v2/ssv/validator/validator.go [hunk 13]
```diff
@@ -89,6 +92,37 @@ func (v *Validator) StartDuty(duty *spectypes.Duty) error {
...
+			if i != 0 {
+				signersStr += "_"
+			}
+			signersStr += fmt.Sprint(signer)
+		}
+		return fmt.Sprintf("%s-%s-%s", msg.MsgID.GetRoleType().String(), msgType, signersStr)
+	}
+	logStart := func() {
+		v.logger.Debug(
+			fmt.Sprintf("started processing %s message", msgType),
```
Comment (lior-blox) [inline protocol/v2/ssv/validator/validator.go:112]: remove

**Review by lior-blox — CHANGES_REQUESTED** [review]

File: protocol/v2/qbft/instance/instance.go (hunk 5 above)
Comment (moshe-blox) [reply protocol/v2/qbft/instance/instance.go]: We returned back to the olds ways, because this broke spectests

**Review by lior-blox — APPROVED** [review]

**Pull request merged**